torch/testing/_internal/inputgen/variable/space.py

```diff
@@ -0,0 +1,233 @@
+            self.values = set()
+        else:
+            if any(isinstance(v, float) and math.isnan(v) for v in values):
+                raise Exception("NaN values are not supported")
+            self.initialized = True
+            self.values = set(values)
+
+    def __str__(self) -> str:
+        if len(self.values) == 0:
+            return "{}"
+        return str(self.values)
+
+    def empty(self) -> bool:
+        if not self.initialized:
+            raise Exception("Discrete must be initialized before checking if empty")
+        return len(self.values) == 0
+
+    def contains(self, v) -> bool:
+        if not self.initialized:
+            raise Exception("Discrete must be initialized before checking membership")
+        return v in self.values
+
+    def remove(self, v) -> None:
+        if not self.initialized:
+            raise Exception("Discrete must be initialized before removing")
+        self.values.difference_update({v})
+
+    def filter(self, f) -> None:
+        if not self.initialized:
+            raise Exception("Discrete must be initialized before filtering")
+        new_values = set()
+        for v in list(self.values):
+            if f(v):
+                new_values.add(v)
+        self.values = new_values
+
+
+class Interval:
+    def __init__(
+        self,
+        lower: Union[int, float] = float("-inf"),
+        upper: Union[int, float] = float("inf"),
+        lower_open: bool = False,
+        upper_open: bool = False,
+    ):
+        self.lower = lower
+        self.upper = upper
+        self.lower_open = lower_open
+        self.upper_open = upper_open
+
+    def __str__(self) -> str:
+        lower_braket = "(" if self.lower_open else "["
+        upper_braket = ")" if self.upper_open else "]"
+        return f"{lower_braket}{self.lower}, {self.upper}{upper_braket}"
+
+    def empty(self) -> bool:
+        if self.lower < self.upper:
+            return False
+        elif self.lower == self.upper:
+            return self.lower_open or self.upper_open
+        else:
+            return True
+
+    def contains(self, v) -> bool:
+        if self.empty():
+            return False
+        if v < self.lower:
+            return False
+        if v == self.lower and self.lower_open:
+            return False
+        if v == self.upper and self.upper_open:
+            return False
+        if v > self.upper:
+            return False
+        return True
+
+    def contains_int(self) -> bool:
+        if self.empty():
+            return False
+        if self.lower > INT64_MAX or (self.lower == INT64_MAX and self.lower_open):
+            return False
+        if self.upper < INT64_MIN or (self.upper == INT64_MIN and self.upper_open):
+            return False
+        if self.upper - self.lower > 1:
+            return True
+        return self.contains(math.ceil(self.lower)) or self.contains(
+            math.floor(self.upper)
+        )
+
+
+class Intervals:
```

Comment:
Can this be a subclass of `Discrete`? A lot of the APIs look the same

Reply:
No. They are different objects.